java-checks/src/main/java/org/sonar/java/checks/security/AndroidPersistentUniqueIdentifierCheck.java

```diff
@@ -0,0 +1,131 @@
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodReferenceTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S7435")
+public class AndroidPersistentUniqueIdentifierCheck extends AbstractMethodDetection {
+
+  private static final String HARDWARE_ID_MESSAGE = "Using a hardware identifier puts user privacy at risk. Make sure it is safe here.";
+  private static final String PHONE_NUMBER_MESSAGE = "Using a phone number puts user privacy at risk. Make sure it is safe here.";
+  private static final String ADVERTISING_ID_MESSAGE = "Using Advertising ID puts user privacy at risk. Make sure it is safe here.";
+  private static final String NON_RESETTABLE_PERSISTENT_ID_MESSAGE = "Using a non-resettable persistent identifier puts user privacy at risk. Make sure it is safe here.";
+
+  private static final MethodMatchers STATIC_SETTINGS_SECURE_GET_STRING_MATCHER =
+    MethodMatchers.create()
+      .ofTypes("android.provider.Settings$Secure")
+      .names("getString")
+      .addParametersMatcher("android.content.ContentResolver", "java.lang.String")
+      .build();
+
+  private static final Map<MethodMatchers, String> MATCHERS = Map.of(
+    MethodMatchers.create()
+      .ofTypes("android.bluetooth.BluetoothAdapter")
+      .names("getAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.net.wifi.WifiInfo")
+      .names("getMacAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getSimSerialNumber", "getDeviceId", "getImei", "getMeid")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getLine1Number")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.SubscriptionManager")
+      .names("getPhoneNumber")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes(
+        "com.google.android.gms.ads.identifier.AdvertisingIdClient$Info",
+        "androidx.ads.identifier.AdvertisingIdInfo",
+        "com.huawei.hms.ads.identifier.AdvertisingIdClient$Info"
+      )
+      .names("getId")
+      .withAnyParameters()
+      .build(),
+    ADVERTISING_ID_MESSAGE,
+    STATIC_SETTINGS_SECURE_GET_STRING_MATCHER,
+    NON_RESETTABLE_PERSISTENT_ID_MESSAGE
+  );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.METHOD_REFERENCE);
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(MATCHERS.keySet().toArray(MethodMatchers[]::new));
```

Comment:
The `getMethodInvocationMatchers` is called by the superclass for every node we visit; I would store the matcher we are creating here somewhere else instead of instantiating it every time, to save up on performance

Reply:
Fair point! I have extracted the matcher construction into a static field.